Better image preparation

baneviciute opened the pull request.

baneviciute committed:
Extracted info from scanned text using regex

baneviciute committed:
Implement formatting of receipts.

baneviciute committed:
minor fixes

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Merge branch 'master' of https://github.com/GREEDY2/GREEDY

baneviciute committed:
Changed the whole image preparation class. Now images are being cropped and filter completely diferent. As a result (for as much testing as i have done) the OCR now reads much better.

baneviciute committed:
Added noise removal from an image

MariusKurbakovas commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+
+namespace GREEDY.ImagePreparation
+{
+    class Deskew
```
Don't know if it is just me but the class name gives no idea what does it do. Also shouldn't it be a public class?

MariusKurbakovas commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,117 @@
+            //Image<Gray, byte> edited = image.SmoothBlur(10, 10, true);
+            Image<Gray, byte> edited = image.SmoothMedian(7);//(15);
+            //Image<Gray, byte> edited = image.SmoothBilatral(7, 255, 34);
+            //Image<Gray, byte> edited = image.SmoothGaussian(3, 3, 34.3, 45.3);
```
What are these commented lines?

valentk777 replied:
yes, you should remove during testing created parts

MariusKurbakovas reviewed (commented):
If this works then very nice. Added a few comments without going deeper in to the code. Also this code can be prone to exceptions, so I think it will need more testing and exception handling in the future

MariusKurbakovas reviewed (commented):
This branch and Valentinas has merge conflicts

valentk777 commented on GREEDY.UnitTests/ImagePreparationUnitTest/UnitTest_ImageFormating.cs:
```diff
@@ -28,6 +30,6 @@ public void ImageFormating_BradleyLocalThreshold_NullBitmap()
             //assert
             Assert.Null(imageFormating.BradleyLocalThreshold(null));
-        }
+        }*/
```
so, you need to replace this unit test with new one or delete it. Do not leave unit test witch is old and not testing our program at all

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+            public double Alpha;
+        }
+
+
```
remove extra lines

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+        }
+
+
+        // The Bitmap
```
too much comments. It is obvious what you do like in this place "Bitmap _internalBmp;". The main thin is to create as more informative (and short) name as possible. this one is good enough. it is not needed additional explanation

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+        // Range of d
+        double _min;
+
+
```
remove extra lines

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+                g.TranslateTransform(-bmp.Width / 2, -bmp.Height / 2);
+                g.DrawImage(bmp, new Point(0, 0));
+            }
+
```
extra lines

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+    class Deskew
+    {
+        // Representation of a line in the image.  
+        private class HougLine
```
class in the class? you should change it :)

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+
+            for (alpha = 0; alpha <= STEPS - 1; alpha++)
+            {
+                double d = y * _cosA[alpha] - x * _sinA[alpha];
```
why you make a declaration every time in the for?

valentk777 commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,117 @@
 
-        //returns the bigest found rectangle
-        private Bitmap FilterCropedImages(List<Bitmap> list)
+        //Not really sure what it does but the recomendations suggest it helps the OCR read better
```
"Tesseract works best on images which have a DPI of at least 300 dpi, so it may be beneficial to resize images" official Tesseract page. so, you have to change comment, because we know that it does and why we need it :)

valentk777 reviewed (changes requested):
all good, but need cosmetic changes.

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+        int[] _hMatrix;
+
+
+        private Bitmap RotateImage(Bitmap bmp, float angle)
```
why you need to rotate image if you did it before? or maybe you can use the same class if you really need to rotate at this point

valentk777 commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,117 @@
-            {
-                return null;
-            }
+            Bitmap rescaledBitmap = new Bitmap(bitmap);
```
why you create a new bitmap when you already get a bitmap?

valentk777 commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,117 @@
+        /// <returns></returns>
+        private Bitmap Rotate(Bitmap bitmap)
+        {
+            Bitmap rotatedBitmap = new Bitmap(bitmap);
```
why you create a new bitmap?

valentk777 commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,117 @@
-            UMat uimage = new UMat();
-            CvInvoke.CvtColor(img, uimage, ColorConversion.Bgr2Gray);
+            //Image<Gray, byte> edited = image.SmoothBlur(10, 10, true);
+            Image<Gray, byte> edited = image.SmoothMedian(7);//(15);
```
magic number 7

valentk777 commented on GREEDY/ImagePreparation/Deskew.cs:
```diff
@@ -0,0 +1,203 @@
+                sum += hl[i].Alpha;
+                count += 1;
+            }
+            return sum / count;
```
why you need count if you calculate 20 times? (i<=19)

Commit:
UnitTest for image processing added
update packages
fix code style a bit (using linq, claning few code lines and others)

AidasKil commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -1,171 +1,91 @@
     {
-        // adds BradleyLocalThreshold filter to the picture
-        public Bitmap BradleyLocalThreshold(Bitmap picture) 
+        private static DeskewImage DeskewImage => new DeskewImage();
```
a new instance will be created each time this property is referenced.

Commit:
small changes for Aidas code review comments
